### PR TITLE
Add RoCBert support for Bettertransformer

### PR DESCRIPTION
**Pull request by shogohida**
# What does this PR do?
Adds RoCBert support for Bettertransformer

Fixes https://github.com/huggingface/transformers/issues/20372

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

### CODE REVIEW

**Comment by shogohida** [comment]
I got the following error but I didn't get it....

https://github.com/huggingface/optimum/actions/runs/3608885843/jobs/6081897113
https://github.com/huggingface/optimum/actions/runs/3608885843/jobs/6081897157
```
 File "/Users/runner/work/optimum/optimum/tests/bettertransformer/testing_bettertransformer_utils.py", line 110, in test_raise_autocast
    _ = bt_model(**inputs)
  File "/Users/runner/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1190, in _call_impl
    return forward_call(*input, **kwargs)
  File "/Users/runner/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/transformers/models/roc_bert/modeling_roc_bert.py", line 1041, in forward
    embedding_output = self.embeddings(
  File "/Users/runner/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1190, in _call_impl
    return forward_call(*input, **kwargs)
  File "/Users/runner/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/transformers/models/roc_bert/modeling_roc_bert.py", line 282, in forward
    embedding_in = self.LayerNorm(embedding_in)
  File "/Users/runner/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1190, in _call_impl
    return forward_call(*input, **kwargs)
  File "/Users/runner/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/torch/nn/modules/normalization.py", line 190, in forward
    return F.layer_norm(
  File "/Users/runner/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/torch/nn/functional.py", line 2515, in layer_norm
    return torch.layer_norm(input, normalized_shape, weight, bias, eps, torch.backends.cudnn.enabled)
RuntimeError: expected scalar type BFloat16 but found Float
```

File: optimum/bettertransformer/models/encoder_models.py [hunk 1]
```diff
@@ -1068,3 +1068,111 @@ def forward(self, hidden_states, attention_mask, position_bias=None, *_, **__):
...
+            self.norm2_bias,
+            self.linear1_weight,
+            self.linear1_bias,
+            self.linear2_weight,
+            self.linear2_bias,
+            attention_mask,
+        )
+        if hidden_states.is_nested and self.is_last_layer:
+            hidden_states = hidden_states.to_padded_tensor(0.0)
+        return (hidden_states,)
```
Comment (younesbelkada) [inline optimum/bettertransformer/models/encoder_models.py:1178]:
```suggestion

```

File: optimum/bettertransformer/models/__init__.py [hunk 2]
```diff
@@ -20,6 +20,7 @@
     DistilBertLayerBetterTransformer,
     FSMTEncoderLayerBetterTransformer,
     MBartEncoderLayerBetterTransformer,
+    RoCBertLayerBetterTransformer,
```
Comment (younesbelkada) [inline optimum/bettertransformer/models/__init__.py:23]:
```suggestion
```

File: optimum/bettertransformer/models/__init__.py [hunk 3]
```diff
@@ -41,6 +42,7 @@
     "LayoutLMLayer": BertLayerBetterTransformer,
     "BertGenerationLayer": BertLayerBetterTransformer,
     "XLMRobertaLayer": BertLayerBetterTransformer,
+    "RoCBertLayer": RoCBertLayerBetterTransformer,
```
Comment (younesbelkada) [inline optimum/bettertransformer/models/__init__.py:45]:
```suggestion
    "RoCBertLayer": BertLayerBetterTransformer,
```

**Review by younesbelkada — COMMENTED** [review]
Thanks so much for adding `BetterTransformer` support for this new architecture!
SInce `ROCBert` copies exactly the same structure as `Bert`, we can leverage that and use directly `BertLayerBetterTransformer` as the converted layer :D 
Regarding the failing test, let me dig a little bit and get back to you

**Comment by younesbelkada** [comment]
Hi @shogohida 
Don't worry about this test as it is flaky ;) Will look into the failing BT test now

**Comment by fxmarty** [comment]
@shogohida Thanks for the work! This is a flaky test, fixed in https://github.com/huggingface/optimum/pull/564 . I rerun the workflow and it should be fine.

**Comment by fxmarty** [comment]
Yet an other test failing:

```
======================================================================
ERROR: test_raise_autocast (test_bettertransformer_encoder.BetterTransformersEncoderTest)
A tests that checks if the conversion raises an error if the model is run under
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/optimum/optimum/tests/bettertransformer/testing_bettertransformer_utils.py", line 141, in test_raise_autocast
    _ = bt_model(**inputs)
  File "/opt/hostedtoolcache/Python/3.8.15/x64/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1190, in _call_impl
    return forward_call(*input, **kwargs)
  File "/opt/hostedtoolcache/Python/3.8.15/x64/lib/python3.8/site-packages/transformers/models/roc_bert/modeling_roc_bert.py", line 1041, in forward
    embedding_output = self.embeddings(
  File "/opt/hostedtoolcache/Python/3.8.15/x64/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1190, in _call_impl
    return forward_call(*input, **kwargs)
  File "/opt/hostedtoolcache/Python/3.8.15/x64/lib/python3.8/site-packages/transformers/models/roc_bert/modeling_roc_bert.py", line 282, in forward
    embedding_in = self.LayerNorm(embedding_in)
  File "/opt/hostedtoolcache/Python/3.8.15/x64/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1190, in _call_impl
    return forward_call(*input, **kwargs)
  File "/opt/hostedtoolcache/Python/3.8.15/x64/lib/python3.8/site-packages/torch/nn/modules/normalization.py", line 190, in forward
    return F.layer_norm(
  File "/opt/hostedtoolcache/Python/3.8.15/x64/lib/python3.8/site-packages/torch/nn/functional.py", line 2515, in layer_norm
    return torch.layer_norm(input, normalized_shape, weight, bias, eps, torch.backends.cudnn.enabled)
RuntimeError: expected scalar type BFloat16 but found Float
```

**Comment by shogohida** [comment]
Thanks for your comments guys! So what should I do?.... Can we skip the failing test as Younes said?

**Comment by younesbelkada** [comment]
Hi @shogohida 
Thanks so much for the heads up!
Yes let's skip the test for now. can you create a separate testing class for ROCBert (check what we have done for `Vilt` for example) Let me know if I can help !

**Comment by shogohida** [comment]
@younesbelkada 
Sure! I need to create a separate test class like [this](https://github.com/huggingface/optimum/pull/508/files#diff-0878efaf9fa78b603dcd1adedac16e3930cd167f37c4018b7b7dc4fbf92c8418R54-R68), right?

I'll let you know if I get stuck!

**Comment by younesbelkada** [comment]
This is correct, please proceed as suggested ;-)

**Comment by shogohida** [comment]
@younesbelkada 
Sorry for coding this late. I added a separate test class for RocBert but I didn't know how to change `prepare_inputs_for_class`... I read other parts of the code but I couldn't find the solution and just copied from `BetterTransformersEncoderTest`.

https://github.com/huggingface/optimum/pull/542/commits/61ff47323010245b0ca986dc0d227e99da80293a

**Comment by fxmarty** [comment]
Hi @shogohida , thanks for working on it! Maybe

```python
class BetterTransformersRoCBertTest(BetterTransformersEncoderTest):
    all_models_to_test = ["path-to-tiny-rocbert-model-here"]
    
    def test_raise_autocast(self):
        pass
```

would work

**Comment by shogohida** [comment]
Hi @fxmarty 
Thanks for your comment! I changed the code as you indicated.

https://github.com/huggingface/optimum/pull/542/commits/228aa1591622ee32600a85794b3c02da31fecd96

**Comment by shogohida** [comment]
Still facing the same error....

https://github.com/huggingface/optimum/actions/runs/3882552677/jobs/6630181175
https://github.com/huggingface/optimum/actions/runs/3882552677/jobs/6630181346

```
Traceback (most recent call last):
  File "/Users/runner/work/optimum/optimum/tests/bettertransformer/testing_bettertransformer_utils.py", line 141, in test_raise_autocast
    _ = bt_model(**inputs)
  File "/Users/runner/hostedtoolcache/Python/3.8.15/x64/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1194, in _call_impl
    return forward_call(*input, **kwargs)
  File "/Users/runner/hostedtoolcache/Python/3.8.15/x64/lib/python3.8/site-packages/transformers/models/roc_bert/modeling_roc_bert.py", line 1041, in forward
    embedding_output = self.embeddings(
  File "/Users/runner/hostedtoolcache/Python/3.8.15/x64/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1194, in _call_impl
    return forward_call(*input, **kwargs)
  File "/Users/runner/hostedtoolcache/Python/3.8.15/x64/lib/python3.8/site-packages/transformers/models/roc_bert/modeling_roc_bert.py", line 282, in forward
    embedding_in = self.LayerNorm(embedding_in)
  File "/Users/runner/hostedtoolcache/Python/3.8.15/x64/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1194, in _call_impl
    return forward_call(*input, **kwargs)
  File "/Users/runner/hostedtoolcache/Python/3.8.15/x64/lib/python3.8/site-packages/torch/nn/modules/normalization.py", line 190, in forward
    return F.layer_norm(
  File "/Users/runner/hostedtoolcache/Python/3.8.15/x64/lib/python3.8/site-packages/torch/nn/functional.py", line 2515, in layer_norm
    return torch.layer_norm(input, normalized_shape, weight, bias, eps, torch.backends.cudnn.enabled)
RuntimeError: expected scalar type BFloat16 but found Float
```

File: tests/bettertransformer/test_bettertransformer_encoder.py [hunk 4]
```diff
@@ -43,6 +43,7 @@
     "hf-internal-testing/tiny-random-MarkupLMModel",
     "hf-internal-testing/tiny-random-rembert",
     "hf-internal-testing/tiny-random-RobertaModel",
+    "hf-internal-testing/tiny-random-RoCBertModel",
```
Comment (fxmarty) [inline tests/bettertransformer/test_bettertransformer_encoder.py:46]: Can you remove rocbert from here?

**Comment by fxmarty** [comment]
Thank you for your contribution!

**Comment by shogohida** [comment]
Thanks for your review! It took time but it was merged in the end... This was my first issue so I hope to contribute more!

**Pull request merged**